rules/mapping.smk

```diff
@@ -1,6 +1,20 @@
+rule bamtofastq:
+    input:
+        bam_file = get_bam
+    output:
+        fastq1 = temp("fastq/{sample}-{unit}_1.fq"),
+        fastq2 = temp("fastq/{sample}-{unit}_2.fq")
+    log:
+        "logs/bamtofastq/{sample}-{unit}.log"
+    threads:
+        4
+    wrapper:
+        get_wrapper_path("bedtools", "bamtofastq")
+
 rule map_reads:
     input:
-        reads = get_fastq
+        reads = get_fastq if config["run"]["bam"] == False else [
```

Comment:
This line means that the pipeline can only handle fastq's or bams as input, but not both right?

Reply:
I think get_fastq is going to be where the magic happens. 

The function should slice on sample name. 

There should only be 3 cases:
1) fq1 is only populated
2) fq1 and fq2 are populated
3) bam is only populated

Otherwise an exception needs to be thrown. I haven't looked into the schema yet, but if it's impossible to implement the sanity check there, then it has to be done in the common.smk code. But hold off on implementing the sanity check for now.

Cases 1 and 2 are already implemented, just return the paths to the fastq files from the TSV file.

Case 3 needs to return the wildcards from bam2fastq (e.g. fastq/{sample}-{unit}_1.fq, fastq/{sample}-{unit}_2.fq)

The pipeline should (fingers crossed) understand that files from case 3 don't exist and can only be produced by bam2fastq. If this works, you can remove the conditional in the config.yaml file and further simplify the pipeline code.

Reply:
Yes, I don't think there'd be any situation in which we have both bams and fastqs. Ah, I see what you're saying with get_fastq returning the input to bamtofastq. I'll look into that.

Reply:
changed this back to just being 'get_fastq' and modified the rule in common.smk